src/jit/decompose.cpp

```diff
@@ -0,0 +1,1189 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+/*XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+XX                                                                           XX
+XX                               Decompose                                   XX
+XX                                                                           XX
+XX Code to decompose 64-bit LONG operations on 32-bit platforms into         XX
+XX multiple single-register operations so individual register usage and      XX
+XX requirements are explicit for LSRA.                                       XX
+XX                                                                           XX
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX
+*/
+
+#include "jitpch.h"
+#ifdef _MSC_VER
+#pragma hdrstop
+#endif
+
+#ifndef LEGACY_BACKEND // This file is ONLY used for the RyuJIT backend that uses the linear scan register allocator
+#ifndef _TARGET_64BIT_ // Decompose is only used on 32-bit platforms
+
+#include "decompose.h"
+
+//------------------------------------------------------------------------
+
+//------------------------------------------------------------------------
+// PrepareForDecomposition: Do one-time preparation required for LONG decomposition.
+// Namely, promote long variables to multi-register structs.
+//
+// Arguments:
+//    None
+//
+// Return Value:
+//    None.
+//
+void Decompose::PrepareForDecomposition()
+{
+    comp->lvaPromoteLongVars();
+}
+
+
+//------------------------------------------------------------------------
+// DecomposeBlock: Do LONG decomposition to all the statements in the given block.
+// This must be done before lowering the block, as decomposition can insert
+// additional statements.
+//
+// Decomposition is done as a post-order tree walk. Lower levels of the tree can
+// create new nodes that need to be further decomposed at higher levels. That is,
+// the decomposition "bubbles up" the tree.
+//
+// Arguments:
+//    block - the block to process
+//
+// Return Value:
+//    None.
+//
+void Decompose::DecomposeBlock(BasicBlock* block)
+{
+    for (GenTree* stmt = block->bbTreeList; stmt != nullptr; stmt = stmt->gtNext)
+    {
+        if (stmt->gtFlags & GTF_STMT_SKIP_LOWER)
```

Comment:
Side note: this is incorrect. Should be fixed when we deal with duplicate decomposition caused by embedded temporary creation logic.